### ISSUE TITLE
TorDNSEL not working?

### ISSUE THREAD

**Issue by BarbzYHOOL** [dapphp/TorUtils#3]
Hello,

I simply copied the torDNSEL file and called it like that:
```
use Dapphp\TorUtils\TorDNSEL;

require_once 'TorDNSEL.php';
```
Then I used this code:
```
try {
    $isTor = TorDNSEL::IpPort(
        $_SERVER['SERVER_ADDR'],
        $_SERVER['SERVER_PORT'],
        $_SERVER['REMOTE_ADDR']
    );
    var_dump($isTor);
 } catch (\Exception $ex) {
     echo $ex->getMessage() . "\n";
 }
```

And whether I'm using tor browser 7.5 or not, it returns "bool(false) " (note that I tried functions from other people and same result basically)

**Comment by dapphp** [comment]
Is your site using accessed through a proxy or accelerator like CloudFlare?  If so you will need to adjust `$_SERVER['REMOTE_ADDR']` to the appropriate value.  That is the only reliable field for getting the client IP, but if you're site is using an nginx/Apache proxy or CDN, it will not reflect the client's IP.

Please test your results against https://drew-phillips.com/ip-info/ (the `TorDNSEL check:` line uses the above code).

For CloudFlare you should use something like: 

    if (SITE_CONFIGURED_FOR_CLOUDFLARE) {
        $remote_addr = $_SERVER['HTTP_CF_CONNECTING_IP'];
    }

Don't use that header if your site isn't using CF since it can be spoofed and mislead you.  Always use REMOTE_ADDR unless you know and trust a different header for reporting the proxied client IP.

**Comment by BarbzYHOOL** [comment]
I don't use a proxy or cloudflare, so what should I do?

No idea how to test anything with https://drew-phillips.com/ip-info/ (though I went there with my tor browser and it detected it perfectly which is cool)

**Comment by dapphp** [comment]
My page uses the exact example code in your first post so you should have similar results.

Is your `SERVER_ADDR` not a public IP and instead private like 10.0.0.0/8 or 192.168.0.0/16?  That would likely cause negative results.

The other thing is to just check the Tor IP you are using when the check says it is not using Tor and look it up at https://metrics.torproject.org/rs.html and make sure that it is listed.  Super new exists might not appear in all directories.

**Comment by BarbzYHOOL** [comment]
> Is your SERVER_ADDR not a public IP and instead private like 10.0.0.0/8 or 192.168.0.0/16? That would likely cause negative results.

Yes, exactly, it's because I'm using virtual machines on my server. What can I do?

**Comment by dapphp** [comment]
In the sample code, change the first 2 arguments, `$_SERVER['SERVER_ADDR']`, `$_SERVER['SERVER_PORT']` to your actual public IP address and port.

You can also spoof it to something that's most likely not blocked by any exit relays, like `TorDNSEL::IpPort('9.9.9.9', 443, $_SERVER['REMOTE_ADDR']);` but I'd highly recommend using your actual public IP and port.

The more privacy-centric approach would be to run a Tor daemon locally (it doesn't have to relay) and use the ControlClient to dump all exit IP's to a file.  This way you don't send DNS requests over the internet to the Tor DNSEL service constantly.  If you do use DNSEL, caching is recommended.

Yet another alternative is to check the IP against a list of IP's using one of the various services that publish such lists.  I offer some here: https://openinternet.io/tor/tor-exit-list.txt and https://openinternet.io/tor/tor-node-list.txt  These are published every 10 minutes.  Since some directories might not yet be in consensus with others, there is a small chance a newly promoted exit may not show in the list for a short period of time, but in my tests I have not ran into it.

**Comment by BarbzYHOOL** [comment]
I don't understand most of the things you said sadly :(

I can't rely on your tor list service because it can go down at any moment forever.

Now, I tried using my real public IP address and it worked perfectly ! Do I need something else ? caching ? or no need with this method ?

And why the hell SERVER_ADDR doesn't show my public IP ?

**Comment by dapphp** [comment]
SERVER_ADDR is whatever your network interface is configured with (`ifconfig`) so it's a private in your setup and the right solution is to change it to your public IP like you did.

If you decide to permanently use TorDNSEL I would *highly* recommend caching.  It is wasteful and will slow everything down if you run the query for the same visitor IP address on every request they make.  Caching the result of an IP address for at least 5 minutes is good.  For example: https://plugins.trac.wordpress.org/browser/vigilantor/tags/1.3.2/vigilantor.php#L413 (doRealTimeLookup).  Setting the timeout fairly low isn't a bad idea in case the Tor DNSEL service goes down for a period of time; then the DNS lookup timeouts won't slow you down too much.

Hope that helps!

**Comment by BarbzYHOOL** [comment]
I only want to use the tordnsel on one page that is basically visited once. I wonder if I should cache anything then.

I looked at your caching code but it's using specific wordpress functions like get_site_transient so I'm lost, once again haha.

PS : yes everything you said helped me already a lot !

**Comment by dapphp** [comment]
Caching seems less necessary if a typical visitor will only land there once.  In the example code the WP `get_site_transient` and `set_site_transient` function would be replaced with your cache get and set functions and the rest doesn't need much of any change.

This conversation gives me an idea about alternative methods using lists.  Maybe soon I'll get around to building a code that can download Tor IP lists from multiple sources where it knows how to parse each one; then you can check from a copy of a list and if one goes down it can use another, or alternate between them.

To summarize the original issue here, since your `SERVER_ADDR` was a private address, the code just needs to be changed slightly to use your public IP instead of a variable:

```
use Dapphp\TorUtils\TorDNSEL;

require_once 'TorDNSEL.php';

try {
    $isTor = TorDNSEL::IpPort(
        'YOUR PUBLIC IP HERE',
        'YOUR SERVER PORT',
        $_SERVER['REMOTE_ADDR'] // client's IP (this is okay)
    );
    var_dump($isTor);
 } catch (\Exception $ex) {
     echo $ex->getMessage() . "\n";
 }

```

**Comment by BarbzYHOOL** [comment]
I didn't need to change the port at all though but yes, that was the solution for the original issue. Let me check for my cache functions though.
